test/evergreen/code_change_report/per_test_code_coverage_report.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python3
+#
+# Public Domain 2014-present MongoDB, Inc.
+# Public Domain 2008-2014 WiredTiger, Inc.
+#
+# This is free and unencumbered software released into the public domain.
+#
+# Anyone is free to copy, modify, publish, use, compile, sell, or
+# distribute this software, either in source code form or as a compiled
+# binary, for any purpose, commercial or non-commercial, and by any
+# means.
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import argparse
+import json
+import os
+from pygit2 import Diff
+from code_change_helpers import diff_to_change_list, read_complexity_data, preprocess_complexity_data
+
+
+# Read a json file into a dict
+def read_json(json_file_path: str) -> dict:
+    with open(json_file_path) as json_file:
+        info = json.load(json_file)
+        return info
+
+
+# Collate the code coverage data from a directory full of build directory copies
+# into a dict. Be careful with memory usage as the resulting dict may be large.
```

Comment:
Great question, and the answer is several Gb of RAM for the current number of tests. The reason for loading them all at once is ease of processing rather than memory optimisation, but if we increase the number of tests significantly then we may need to develop a more memory efficient approach. 

My original implementation compiled the all-in-one dictionary in memory and then wrote it as a single json file, but converting the dict to json ran out of memory on the test machine which has 32Gb(!), so I dropped that approach.